### PR TITLE
implement np.convolve method in lax_numpy

### PR DESCRIPTION
**Pull request by michiboo**
implement np.convolve method as described in #1561

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3089,6 +3089,18 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
   return true_divide(dot(X, X_T.conj()), f).squeeze()
 
 
+@_wraps(onp.convolve)
+def convolve(x, y, mode='full'):
+  x_jax = lax.reshape(x,(1,1,len(x)))
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Minor style nit: elsewhere in JAX, the style is to put a space after most commas, so here:
` x_jax = lax.reshape(x, (1, 1, len(x)))`
Same everywhere.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3089,6 +3089,18 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
...
+  y_jax = flip(lax.reshape(y,(1,1,len(y))),2)
+  if mode == 'full':
+    return ravel(lax.conv_general_dilated(x_jax,y_jax,[1],[(len(x)-1,len(x)-1)]))
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Is this the right padding? I don't think so: I would have assumed the padding would be related to the size of `y`, not the size of `x`. Certainly this doesn't seem to produce the right output when `x` and `y` don't have the same size.

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -2119,6 +2119,19 @@ def testCov(self, shape, dtype, rowvar, ddof, bias, rng_factory):
     self._CompileAndCheck(lnp_fun, args_maker, check_dtypes=True, atol=tol,
                           rtol=tol)
 
+  def test_convolve(self):
+    x = onp.array([1.,2.,3.])
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py]: Can you test multiple sizes of `x` and `y`? You would probably do well to make this a parametric test, as many others in this file are.

You could also avoid redundancy in the test by making the `mode` argument parametric as well.

File: jax/numpy/lax_numpy.py [hunk 4]
```diff
@@ -3089,6 +3089,18 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
   return true_divide(dot(X, X_T.conj()), f).squeeze()
 
 
+@_wraps(onp.convolve)
+def convolve(x, y, mode='full'):
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: I suspect you need some type promotion logic here. For example, does this implementation work similarly to:
```
onp.convolve(onp.arange(5), onp.arange(3, dtype=onp.float32))
```

Most likely what you need is the line:
```
x, y = _promote_types(x, y)
```

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by michiboo** [comment]
@hawkinsp I have updated the code, also I made some changes to lax.padtype_to_pads, where it cannot handle cases where the window shape is bigger than in_shape

**Comment by hawkinsp** [comment]
Did you push the changes? I don't see them in the PR yet.

**Comment by michiboo** [comment]
@hawkinsp  sorry, I type git *.py and it didn't add to the new commits. Anyway there is apparently some changes that made the CI test failed. As you can see I commented out all my changes and test and CI still failed. I tested on my local machine and the tests all passed

**Comment by michiboo** [comment]
@hawkinsp Hi I don't understand why my test failed on CI, it passed on my local machine.

**Comment by michiboo** [comment]
@hawkinsp hi, finally the CI test passed, there is a off by one error compare to numpy's result when padding=same and lhr is even number. it would be great if you can review it

File: jax/lax/lax.py [hunk 5]
```diff
@@ -1126,7 +1126,7 @@ def _safe_mul(x, y): return safe_mul_p.bind(x, y)
 ### convenience wrappers around traceables
 
 
-def conv(lhs, rhs, window_strides, padding, precision=None):
+def conv(lhs, rhs, window_strides, padding, precision=None, convolve=False):
```
Comment (mattjj) [inline jax/lax/lax.py]: Can we avoid the modification to `lax.py`? Perhaps you can just call `lax.conv_general_dilated` directly from the `np.convolve` implementation, and basically move these lines over there.

File: jax/numpy/lax_numpy.py [hunk 6]
```diff
@@ -3088,6 +3088,19 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
   X_T = X.T if w is None else (X * w).T
   return true_divide(dot(X, X_T.conj()), f).squeeze()
 
+@_wraps(onp.convolve)
+def convolve(x, y, mode='full'):
+  if shape(y)[0] > shape(x)[0]:
```
Comment (mattjj) [inline jax/numpy/lax_numpy.py]: It might be a good idea to check that the inputs are 1D, and raise an error otherwise.

**Comment by mattjj** [comment]
Hey @michiboo , sorry for the delay on our end (from NeurIPS followed by holidays).

This looks like it's in pretty good shape! I left a couple small comments just now about tweaks.

Just to clarify from your last comment:

> there is a off by one error compare to numpy's result when padding=same and lhr is even number. it would be great if you can review it

Do you mean that the current implementation has a known bug in it, or that you have already fixed this problem?

**Comment by michiboo** [comment]
> Hey @michiboo , sorry for the delay on our end (from NeurIPS followed by holidays).
> 
> This looks like it's in pretty good shape! I left a couple small comments just now about tweaks.
> 
> Just to clarify from your last comment:
> 
> > there is a off by one error compare to numpy's result when padding=same and lhr is even number. it would be great if you can review it
> 
> Do you mean that the current implementation has a known bug in it, or that you have already fixed this problem?

Hi, it is probably due to the fact that the Conv method in jax is implemented similar to tensorflow while the convolve implemented in numpy use its own conv method. 
I already fixed the problem by offset it by one.

**Comment by michiboo** [comment]
bump

File: jax/numpy/lax_numpy.py [hunk 7]
```diff
@@ -3088,6 +3088,25 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
...
+    x, y = y, x
+  x, y = _promote_dtypes(x, y)
+  x_jax = lax.reshape(x, (1, 1, len(x)))
+  y_jax = flip(lax.reshape(y, (1, 1, len(y))))
+  if mode.lower() == 'full':
+    return ravel(lax.conv_general_dilated(x_jax, y_jax, [1], [(len(y)-1, len(y)-1)]))
+  elif mode.lower() == 'valid':
+    return ravel(lax.conv(x_jax, y_jax, (1,), "VALID"))
+  elif mode.lower() == 'same':
+    pads = lax.padtype_to_pads(x_jax.shape[2:], y_jax.shape[2:], (1,), "SAME")
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:3106]: Could we express this more directly? I think calling `padtype_to_pads` is actually just making the code more complicated, not less. Here I think you just want:
```
low = (len(y) + 1) // 2
high = len(y) - low
```

(although you'd need to double check that, I didn't test that.)

File: tests/lax_numpy_test.py [hunk 8]
```diff
@@ -2119,6 +2119,28 @@ def testCov(self, shape, dtype, rowvar, ddof, bias, rng_factory):
...
 
+  @parameterized.named_parameters(
+      jtu.cases_from_list(
+        {"testcase_name": "_shape={}_shape2={}_dtype={}_mode={}".format(
+            shape, shape2, dtype, mode),
+         "shape": shape, "dtype": dtype, "mode": mode,
+         "shape2": shape2,"rng_factory": rng_factory}
+        for shape in [(i,) for i in range(2, 100)]
+        for shape2 in [(i,) for i in range(2, 100)]
+        for dtype in [onp.float16, onp.float32, onp.float64]
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py:2130]: I'd use `float_dtypes` here rather than enumerating a list of types. Not all hardware platforms support all of these types (e.g., TPU doesn't support `float16`.)

File: tests/lax_numpy_test.py [hunk 9]
```diff
@@ -2119,6 +2119,28 @@ def testCov(self, shape, dtype, rowvar, ddof, bias, rng_factory):
     self._CompileAndCheck(lnp_fun, args_maker, check_dtypes=True, atol=tol,
                           rtol=tol)
 
+  @parameterized.named_parameters(
+      jtu.cases_from_list(
+        {"testcase_name": "_shape={}_shape2={}_dtype={}_mode={}".format(
+            shape, shape2, dtype, mode),
+         "shape": shape, "dtype": dtype, "mode": mode,
+         "shape2": shape2,"rng_factory": rng_factory}
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py:2127]: Space after `shape2,`

File: tests/lax_numpy_test.py [hunk 10]
```diff
@@ -2119,6 +2119,28 @@ def testCov(self, shape, dtype, rowvar, ddof, bias, rng_factory):
...
+        for shape in [(i,) for i in range(2, 100)]
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py:2128]: This generates rather a lot of test cases (~98^2). That's too many. I think you can achieve what you want by just handwriting a few special cases, e.g.,
`[2, 3, 5, 10]`

Does it make sense to test `0` and `1` for this function? Does numpy work for those cases? If so, I'd test them too.

File: jax/lax/lax.py [hunk 11]
```diff
@@ -1142,7 +1142,7 @@ def conv(lhs, rhs, window_strides, padding, precision=None):
     An array containing the convolution result.
   """
   pads = padtype_to_pads(lhs.shape[2:], rhs.shape[2:], window_strides, padding)
-  return conv_general_dilated(lhs, rhs, window_strides, padding,
+  return conv_general_dilated(lhs, rhs, window_strides, pads,
```
Comment (hawkinsp) [inline jax/lax/lax.py:1145]: Revert this change? It looks unintentional and would break this function.

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks for the PR, sorry for the slow review.

**Comment by mattjj** [comment]
I think #2675 built on this work and finished off the underlying issue!
